crates/red_knot_python_semantic/resources/mdtest/annotations/deferred.md

```diff
@@ -45,3 +45,40 @@ class Foo: ...
 
 reveal_type(get_foo())  # revealed: Foo
 ```
+
+## Deferred self-reference annotations in a class definition
+
+```py
+from __future__ import annotations
+
+class Bar:
+    this: Bar
+
+    def f(self: "Bar"):
+        reveal_type(self)  # revealed: Bar
+
+    def g(self: Bar):
+        reveal_type(self)  # revealed: Bar
```

Comment:
I'd feel a bit more comfortable if we used non-`self` parameters in these assertions, just because the questions "What is the type of `self`?" and "What are the legal annotations to apply to `self`?" are both somewhat complex questions. Both are irrelevant questions to the thing that's being tested here, and it's useful to make it clear that this test doesn't have any bearing on those questions, I think:

```suggestion
    def f(self, x: "Bar"):
        reveal_type(self)  # revealed: Bar

    def g(self, x: Bar):
        reveal_type(self)  # revealed: Bar
```